### ISSUE TITLE
Alternative to torch.select_mask

### ISSUE THREAD

**Issue by ttdd11** [pytorch/xla#8968]
## ❓ Questions and Help

Most of the time we can adapted routines to avoid graph recompilations, however there is instance where this is a bit tricky. 

When computing a masked mean, we are currently using sum and valids as follows:

```
replaced = input_tensor*is_valid
sum_valid = replaced.sum()
n_valid = is_valid.sum(dtype=input_tensor.dtype)
return torch.nan_to_num(sum_valid / n_valid)
```

Where valid is 0 or 1 if the entry is in the dataset. 

This effectively calculates the mean while ignoring zeros. This works well when the data  is close to full in most examples. However in some instance we insert sparse data that have a reduced is_valid that is consistent across the data set. The result is that the 0 entries in the is_valid are reinforced, and when testing against the test of sparse data, it won't predict the other entries. 

To avoid this - we have historically used torch.select_mask, which only selects the non-zero entries and the gradients only back prop. through those - basically we don't get the reinforced 0-0. 

I'm wondering if there is an alternative or work around to torch.select_mask as this increases computation time by ~6X because of the frequent recompilations.

Thank you again for this awesome tool and let me know if you have any questions.

**Comment by qihqi** [comment]
> The result is that the 0 entries in the is_valid are reinforced, and when testing against the test of sparse data, it won't predict the other entries.

Let me rephrase the question to see if I understood correctly:

So the above math, and the following math:

```
valid = torch.masked_select(input_tensor, is_valid)
return valid.mean()
```

is the same, but their gradient is different?

I guess, if the goal is to make sure that the gradient of the positions of `input_tensor` where `is_valid ==0` remains 0, then, one could reset it before doing optimizer step?

```
...
loss.backward() # computes gradient
input_tensor.grad = input_tensor.grad * is_valid
optimizer.update()
```
?

**Comment by ttdd11** [comment]
@qihqi i think this could work. I will circle back and see if this does the trick. Thank you sincerely for the help!

**Comment by qihqi** [comment]
The shape of `input_tensor.grad` will be equal to the same of `input_tensor` so there are no way around this.*

*: One way around this is to NOT compute input_tensor.grad, but instead, compute the gradients to the non-zero items and do the insertion manually. Once that works, then you can wrap that logic into https://pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd; but, if you do that, then you likely run into recompilation issues if the number of selected element is different everytime. So there need to be some upper bound / padding strategy. So there are definitely trade-offs to be made.

the extra gradient update `input_tensor.grad = input_tensor.grad * is_valid` should not produce new intermediaries if its all part of the XLA graph (so it will be a temporary from XLA's perpective, but it likely the buffer will be reused by the XLA compiler). You need to try to verify though.

**Comment by ttdd11** [comment]
@qihqi I had another thought that may work - can we just zero the grads on the output or does it have to be the input? 

Ie. if we set the grads on the output to be zero would that prevent the update or would it need to be on the input tensors?

**Comment by qihqi** [comment]
So when you do `loss.backward()` it will compute gradients in order: 
* compute grad of loss, then
* compute grad of nan_to_sum(sum_valid / n_valid), then
* compute the grad sum_valid / n_valid, then
* ...
on this order. 

So, what you want to do, is to let .backward do part of the work, you go in modify the gradient, then let .backward continue.

The way that I know to intercept in the middle of the backward computation chain is via  [pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd](https://pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd); so if you are doing that, then do it at the mask application level.

i.e. make a function called `apply_mask(inputs, masks)` with custom backward. And in that backward you dont materialize the whole gradient.

**Comment by ttdd11** [comment]
@qihqi this is great - thank you very much. I will implement this tomorrow for that and see how that goes. 

Thanks again. I will close for now and if I run into issue with this I'll re-open.

**Comment by ttdd11** [comment]
@qihqi Thanks again for the help on this! I went through that document and based on that I believe the recommendation for this scenario is to register the backward hook and modify the input gradients there. Does that make sense?

It requires a few changes to the forward and ensuring that the functions are modules, but other than that I don't see this being a big lift at all.

**Comment by qihqi** [comment]
> Thanks again for the help on this! I went through that document and based on that I believe the recommendation for this scenario is to register the backward hook and modify the input gradients there. Does that make sense? 

Yup.

> It requires a few changes to the forward and ensuring that the functions are modules, but other than that I don't see this being a big lift at all.

Good to hear, I'll close this for now for my own tracking. Please reopen if you have further questions and/or AIs for us to do. Thanks!
